ipld/resolver/src/observe.rs

```diff
@@ -0,0 +1,312 @@
+        register_int_counter!("membership_publish_failure", "Number of failed publish attempts");
+
+    CONTENT_RESOLVE_RUNNING: IntGauge =
+        register_int_gauge!("content_resolve_running", "Number of currently running content resolutions");
+
+    CONTENT_RESOLVE_NO_PEERS: IntCounter =
+        register_int_counter!("content_resolve_no_peers", "Number of resolutions with no known peers");
+
+    CONTENT_RESOLVE_SUCCESS: IntCounter =
+        register_int_counter!("content_resolve_success", "Number of successful resolutions");
+
+    CONTENT_RESOLVE_FAILURE: IntCounter =
+        register_int_counter!("content_resolve_failure", "Number of failed resolutions");
+
+    CONTENT_RESOLVE_FALLBACK: IntCounter =
+        register_int_counter!("content_resolve_fallback", "Number of resolutions that fall back on secondary peers");
+
+    CONTENT_RESOLVE_PEERS: Histogram =
+        register_histogram!("content_resolve_peers", "Number of peers found for resolution from a subnet");
+
+    CONTENT_CONNECTED_PEERS: Histogram =
+        register_histogram!("content_connected_peers", "Number of connected peers in a resolution");
+
+    CONTENT_RATE_LIMITED: IntCounter =
+        register_int_counter!("content_rate_limited", "Number of rate limited requests");
+}
+
+impl_traceables!(TraceLevel::Info, "Ping", PingEvent);
+impl_traceables!(TraceLevel::Warn, "Ping", PingFailureEvent);
+impl_traceables!(TraceLevel::Info, "Identify", IdentifyEvent);
+impl_traceables!(TraceLevel::Warn, "Identify", IdentifyFailureEvent);
+impl_traceables!(TraceLevel::Info, "Discovery", DiscoveryEvent);
+impl_traceables!(TraceLevel::Info, "Membership", MembershipEvent);
+impl_traceables!(TraceLevel::Warn, "Membership", MembershipFailureEvent);
+impl_traceables!(TraceLevel::Info, "Content", ResolveEvent);
+impl_traceables!(TraceLevel::Warn, "Content", ResolveFailureEvent);
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum PingEvent {
+    Success(PeerId, Duration),
+}
+
+impl Recordable for PingEvent {
+    fn record_metrics(&self) {
+        match self {
+            Self::Success(_, rtt) => {
+                PING_SUCCESS.inc();
+                PING_RTT.observe(rtt.as_millis() as f64);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum PingFailureEvent {
+    Timeout(PeerId),
+    Failure(PeerId, Box<dyn Error>),
+}
+
+impl Recordable for PingFailureEvent {
+    fn record_metrics(&self) {
+        match self {
+            Self::Failure(_, _) => PING_FAILURE.inc(),
+            Self::Timeout(_) => PING_TIMEOUT.inc(),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum IdentifyEvent {
+    Received(PeerId),
+}
+
+impl Recordable for IdentifyEvent {
+    fn record_metrics(&self) {
+        match self {
+            Self::Received(_) => IDENTIFY_RECEIVED.inc(),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum IdentifyFailureEvent {
+    Failure(PeerId, Box<dyn Error>),
```

Comment:
Do we need to store the `Box<dyn Error>`? We don't really need for anything so I wonder of just a string would suffice?

Reply:
Yes, it's not needed, I mainly used it to signal that's it's an error, since I used tuples instead of structs. 
Fixed here: https://github.com/consensus-shipyard/ipc/pull/1207/commits/915cd016c13cee6fa58beeb8a77a73bb6d275c5b